source/common/unified_memory_allocation/src/memory_pool.c

```diff
@@ -6,43 +6,101 @@
  *
  */
 
+#include "memory_tracker_internal.h"
 #include <uma/memory_pool.h>
+#include <uma/memory_pool_ops.h>
 
 #include <assert.h>
 #include <stdlib.h>
 
 struct uma_memory_pool_t {
-    struct uma_memory_pool_ops_t ops;
     void *pool_priv;
+
+    struct uma_memory_pool_ops_t ops;
+
+    struct uma_memory_provider_desc_t *providers;
+    size_t numProviders;
 };
 
-enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, void *params,
-                                uma_memory_pool_handle_t *hPool) {
+enum uma_result_t umaPoolCreate(struct uma_memory_pool_ops_t *ops, struct uma_memory_provider_desc_t *providers,
+                                size_t numProviders, void *params, uma_memory_pool_handle_t *hPool) {
+    if (providers && !numProviders) {
+        return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    if (!providers && numProviders) {
+        return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+    }
+
+    for (size_t i = 0; i < numProviders; i++) {
+        // UMA_MEMORY_PROVIDER_TYPE_TRACKER can only be created by the UMA library
+        if (providers[i].providerType == UMA_MEMORY_PROVIDER_TYPE_TRACKER) {
+            return UMA_RESULT_ERROR_INVALID_ARGUMENT;
+        }
+    }
```

Comment:
if the `UMA_MEMORY_PROVIDER_TYPE_TRACKER` type were not exposed to the user we would not need this check.

Reply:
Yes, that's true. So we probably need to decide whether 'no memory provider' use case is something we want to support.